AppDelegate.mm

```diff
@@ -160,43 +164,51 @@ - (void)applicationDidFinishLaunching:(NSNotification*)notification {
               if (task.terminationStatus != 0) {
                 [fileManager removeItemAtPath:tempDir error:nil];
 
-                dispatch_async(dispatch_get_main_queue(), ^{
-                  showErrorModal(
-                      [NSString stringWithFormat:@"Failed to extract: %i", task.terminationStatus]);
-                });
+                // If extraction fails, it's likely to be a permission issue let's rerun the program
+                // as root
+
+                if (geteuid() != 0) {
+                  STPrivilegedTask* task = [STPrivilegedTask new];
+                  [task setLaunchPath:[NSBundle.mainBundle executablePath]];
+                  [task launch];
+                } else {
+                  dispatch_async(dispatch_get_main_queue(), ^{
+                    showErrorModal([NSString
+                        stringWithFormat:@"Failed to extract: %i", task.terminationStatus]);
+                  });
+                }
                 return;
               }
 
+              auto* sourcePath = [tempDir
+                  stringByAppendingPathComponent:[NSString
+                                                     stringWithFormat:@"%@.app", targetAppName]];
+              auto* targetPath = NSBundle.mainBundle.bundlePath;
               // Rename the final bundle in the temp directory to the target directory.
               // Lets first try using the rename() system call because it can do that
               // atomically even if the the target already exists.
               if (rename(sourcePath.fileSystemRepresentation,
                          targetPath.fileSystemRepresentation) != 0) {
-                // If rename() failed, try to do this in two steps.
-                NSError* moveError = nil;
-                if (([fileManager fileExistsAtPath:targetPath] &&
-                     [fileManager removeItemAtPath:targetPath error:&moveError] != YES) ||
-                    [fileManager moveItemAtPath:sourcePath toPath:targetPath
```

Comment:
@poiru Do we know any cases where this has worked and the `rename` hasn't?

Reply:
Not sure! I think `rename()` doesn't work properly across volumes, for example, but with your changes it would no longer happen.